### PR TITLE
[Feature] Limit frequency of query and load in system level

### LINKED ISSUE #7129: [Feature] Limit cluster resource usage in user granularity
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

In productive environment, the Doris cluster is often facing pressure from many aspects (mainly from stream load and query), cause many resource shortage problem like OOM, especially in shared cluster.
image
As above picture shows, the memory usage waves too big.
I think it's better to have a way to limit the resource usage of each user. Maybe limit the usage frequency is a proper way.

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by MorningLight5**
## Proposed changes

Fix #7129 

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)
- [ ] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

- [x] I have created an issue on (Fix #ISSUE) and described the bug/feature there in detail
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

### CODE REVIEW

**Comment by xinyiZzz** [comment]
good idea,
If FE can get the memory consumption of each query, we can limit the memory used by users, which is what I plan to do in the future.

The actual AdmissionController of Impala does a similar thing, refer to: https://shimo.im/docs/6qxjctpyDHJgPwtw

In #7198, I have counted the real memory consumption of each Query, which may be implemented based on this PR later

**Comment by morningman** [comment]
Hi @tianhui5 , thanks for your contribution.
But better to add document to explain the usage or best practice.
Otherwise, it is hard to review.

**Comment by tianhui5** [comment]
> Hi @tianhui5 , thanks for your contribution. But better to add document to explain the usage or best practice. Otherwise, it is hard to review.

Thanks for remind me. I've add relative doc and unit test, please review this PR

**Comment by tianhui5** [comment]
> Good idea, If FE can get the memory consumption of each Query and Load, we can limit the memory used by users, which is what I plan to do in the future.
> 
> Impala’s AdmissionController does a similar thing, Introduction is here https://shimo.im/docs/6qxjctpyDHJgPwtw
> 
> In #7198, I have counted the real memory consumption of each Query, which may be implemented based on your PR later

Nice work.

File: fe/fe-core/src/main/java/org/apache/doris/system/RateLimiterMgr.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+/**
+ * RateLimiterMgr collect all rate limiter of the system.
+ */
+public class RateLimiterMgr {
+    private RateLimiter systemQPSLimiter;
+    private RateLimiter systemStreamLoadLimiter;
+    private Map<String, RateLimiter> userToQPSLimiter;
+    private Map<String, RateLimiter> userToStreamLoadLimiter;
+
+    public RateLimiterMgr(double QPS, double streamLoadPS) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/system/RateLimiterMgr.java:37]: It is not recommended to pass these values through the constructor's parameters. If we subsequently add more limiter, there will be more arguments to the constructor.

File: fe/fe-core/src/main/java/org/apache/doris/system/RateLimiterMgr.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        this.systemQPSLimiter = RateLimiter.create(QPS);
+        this.systemStreamLoadLimiter = RateLimiter.create(streamLoadPS);
+        this.userToQPSLimiter = Maps.newConcurrentMap();
+        this.userToStreamLoadLimiter = Maps.newConcurrentMap();
+    }
+
+    public synchronized boolean checkRateLimiter(TLimiterType type, String user) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/system/RateLimiterMgr.java:44]: why using `synchronized`? the RateLimiter is thread safe

**Comment by morningman** [comment]
There are two main issues

1. all query requests need to send an RPC to Master FE, which is too much overhead.
2. Limiter is not set in the right place. For example, for query, it should be handled at the place where the query plan is finally sent to BE. For load, the frequency of transactions should be limited.

**Comment by tianhui5** [comment]
> There are two main issues
> 
> 1. all query requests need to send an RPC to Master FE, which is too much overhead.
> 2. Limiter is not set in the right place. For example, for query, it should be handled at the place where the query plan is finally sent to BE. For load, the frequency of transactions should be limited.

Yes, that seems to have too much RPC. I have rewrote the feature with statistic sychronize based on bdbje. Because I think it's fine to use bdbje to store some temperory statistics, and there is too much work to develop a middleware myself.

Please review this patch, thanks for your time!

**Comment by morningman** [comment]
Hi @tianhui5 

Although we had some private discussions about this feature, we eventually needed a complete design and usage document to help all developers and users understand how this feature works and how to use it.

It is recommended that the final design be fully described in the issue #7129  (and the previous design be marked as Deprecated).

Also, I suggest a separate user document describing the functionality from a usage perspective, including the followings:

1. the purpose of the feature (Motivation)
2. what scenarios the feature is applicable to, how to use it, and what impact it will have.
3. whether there are shortcomings for further improvement.

In the previous Doris development, we also had a lot of features that were difficult to maintain due to lack of documentation, and I hope to be able to improve the process in this area in future projects. Thanks!

**Comment by tianhui5** [comment]
> 

Done.

**Comment by tianhui5** [comment]
> Hi @tianhui5
> 
> Although we had some private discussions about this feature, we eventually needed a complete design and usage document to help all developers and users understand how this feature works and how to use it.
> 
> It is recommended that the final design be fully described in the issue #7129 (and the previous design be marked as Deprecated).
> 
> Also, I suggest a separate user document describing the functionality from a usage perspective, including the followings:
> 
> 1. the purpose of the feature (Motivation)
> 2. what scenarios the feature is applicable to, how to use it, and what impact it will have.
> 3. whether there are shortcomings for further improvement.
> 
> In the previous Doris development, we also had a lot of features that were difficult to maintain due to lack of documentation, and I hope to be able to improve the process in this area in future projects. Thanks!
Done.

**Comment by tianhui5** [comment]
The local current query number is reported to Master every certain time, and then send to all FE.
image

**Pull request closed without merging**